core/src/main/scala/kafka/server/BrokerMetadataEventManager.scala

```diff
@@ -0,0 +1,198 @@
+
+  import BrokerMetadataEventManager._
+
+  val queue = new LinkedBlockingQueue[QueuedEvent]
+
+  val thread = new BrokerMetadataEventThread(BrokerMetadataEventThreadName)
+
+  val processor: MetadataEventProcessor = overrideProcessor.getOrElse(this)
+
+  // metrics
+  private val eventQueueTimeHist = newHistogram(EventQueueTimeMetricName)
+  newGauge(EventQueueSizeMetricName, () => queue.size)
+
+  def start(): Unit = {
+    put(StartupEvent)
+    thread.start()
+  }
+
+  def close(): Unit = {
+    try {
+      thread.initiateShutdown()
+      put(ShutdownEvent) // wake up the thread in case it is blocked on queue.take()
+      thread.awaitShutdown()
+    } finally {
+      removeMetric(EventQueueTimeMetricName)
+      removeMetric(EventQueueSizeMetricName)
+    }
+  }
+
+  def put(event: Event): QueuedEvent = {
+    val queuedEvent = new QueuedEvent(event, time.milliseconds())
+    queue.put(queuedEvent)
+    queuedEvent
+  }
+
+  private def pollFromEventQueue(): QueuedEvent = {
+    val hasRecordedValue = eventQueueTimeHist.count() > 0
+    if (hasRecordedValue) {
+      val event = queue.poll(eventQueueTimeTimeoutMs, TimeUnit.MILLISECONDS)
+      if (event == null) {
+        eventQueueTimeHist.clear()
+        queue.take()
+      } else {
+        event
+      }
+    } else {
+      queue.take()
+    }
+  }
+
+  override def processStartup(): Unit = {
+  }
+
+  override def process(metadataMessage: RaftMessage): Unit = {
+    try {
+      trace(s"Handling metadata message:$metadataMessage")
+      val data = metadataMessage.data
+      data match {
+        case brokerRecord: BrokerRecord => handleBrokerRecord(brokerRecord)
+//        case fenceBrokerRecord: FenceBrokerRecord => handleFenceBrokerRecord(fenceBrokerRecord)
+        case topicRecord: TopicRecord => handleTopicRecord(topicRecord)
+//        case removeTopicRecord: RemoveTopicRecord => handleRemoveTopicRecord(removeTopicRecord)
+        case partitionRecord: PartitionRecord => handlePartitionRecord(partitionRecord)
+        case configRecord: ConfigRecord => handleConfigRecord(configRecord)
+        case isrChangeRecord: IsrChangeRecord => handleIsrChangeRecord(isrChangeRecord)
+        case accessControlRecord: AccessControlRecord => handleAccessControlRecord(accessControlRecord)
+//        case delegationTokenRecord: DelegationTokenRecord => handleDelegationTokenRecord(delegationTokenRecord)
+//        case scramRecord: ScramRecord => handleScramRecord(scramRecord)
+//        case featureLevelRecord: FeatureLevelRecord => handleFeatureLevelRecord(featureLevelRecord)
+        case _ => handleUnexpected(metadataMessage)
+      }
+    } catch {
+      case e: FatalExitError => throw e
+      case e: Exception => handleError(metadataMessage, e)
+    }
+  }
+
+  private def handleBrokerRecord(brokerRecord: BrokerRecord): Unit = {
+    replicaManager.maybeUpdateMetadataCache(brokerRecord)
```

Comment:
it's not "maybe" since all events apply to us, and the metadata log is ordered-- we can't see an event twice